lib/puppet/transaction/report.rb

```diff
@@ -386,6 +394,101 @@ def self.default_format
+  def process_resource_inferences
+    persistence = Puppet::Transaction::Persistence.new
+    persistence.load
+
+    new_resources = {}
+
+    resource_statuses.each do |name, status|
+      # Initialize data structure for resource
+      new_resources[name] ||= {}
+      params = new_resources[name]["parameters"] ||= {}
+
+      # Populate new hash with events mapped to params
+      param_to_event = {}
+      status.events.each do |ev|
+        param_to_event[ev.property] = ev
+      end
+
+      status.real_resource.parameters.each do |pname,param|
+        # Skip if if the data is sensitive, or if its just a property
+        next if !param.is_a?(Puppet::Property) || param.sensitive
+
+        # Create new resource parameter definition in new hash
+        new_param_values = params[pname.to_s] ||= {}
+        event = param_to_event[pname.to_s]
+
+        # Retrieve the old system value
+        old_system_value = persistence.get_system_value(name, pname.to_s)
+
+        # Store the new system value for persistence
+        new_param_values["system_value"] = new_system_value(param, event, old_system_value)
+
+        # Calculate corrective change if there is a related event
+        event.corrective_change = is_corrective_change(param, event, old_system_value) if event
+      end
+
+      # Mark resource and report with corrective_change if there is one
+      status.events.each do |e|
+        if e.corrective_change
+          status.corrective_change = true
+          @corrective_change = true
+          break
+        end
+      end
+    end
+
+    # Persist our new values to the resources storage.
+    persistence.resources = new_resources
+    persistence.save
+  end
+
+  # Given an event and its property, calculate the system_value to persist
+  # for future calculations.
+  # @param [Puppet::Transaction::Event] event event to use for processing
+  # @param [Puppet::Property] property correlating property
+  # @param [Object] system_value system_value from last transaction
+  # @return [Object] system_value to be used for next transaction
+  def new_system_value(property, event, system_value)
+    if event
+      if event.status == "success"
+        # Success events, we presume work so we persist the desired_value
+        event.desired_value
+      else
+        # For non-success events, we persist the old system_value if it exists,
+        # or use the event previous_value.
+        system_value.nil? ? event.previous_value : system_value
+      end
+    else
+      # For non events, we just want to store the parameters agent value.
+      property.value
+    end
+  end
+
+  # @param [Puppet::Transaction::Event] event event to use for processing
+  # @param [Puppet::Property] property correlating property
+  # @param [Object] system_value system_value from last transaction
+  # @return [bool] system_value to persist
+  def is_corrective_change(property, event, system_value)
```

Comment:
So last weekend and earlier this week I attempted to do the calculation for the new system_value in ResourceHarness and the corrective_change calculation in Event. I must admit, it became a complex fairly quickly.

Let me elaborate ... (it will help me think, bear with me :-)

* I need to provide a path back to the Persistence object that we create, this means that I need to have paths from Event & ResourceHarness back to the Transaction to obtain the object reference. There are other links that are missing that are required for this also. Not impossible, but requires work.
* The Persistence mechanism needs to be changed, so that old values and new values for resources are accessible separately. The calculation for the new system_value will need to store the value, while in another place we still need to retrieve the old value for corrective_change calculation. Even with correct ordering (which can be tricky) whereby we replace old with new in the same pass, this leaves us with dangling/orphaned data at the end, as we aren't doing a full flush of old data, and we're back to the state.yaml problem. Without something like this, we're also back to forcing the storing/calculation in one inherit place. I don't think this is impossible, its just not a pattern I'm comfortable with, but perhaps there is another approach, or I can learn to like it.
* Parts of the inner loop (ResourceHarness for example) work in ways to solve the problem they were originally assigned. Trying to follow the same paths as I have on the outer loop (that is, the output) isn't as clear and I found myself with a harder state to debug. This is largely because the code never has had a demand to do what I'm doing.

Mostly I'm afraid this might mean some structural changes or scary changes. At least thats where my patch ended up, I wasn't impressed with my attempt anyway, and more so I didn't get it working. To be straight, this implementation was easier.

Having said that, I feel you are right, but not for the reasons you have listed per se, I think more because if the core of Puppet isn't aligned for this use-case then it possible should be. I think this would aid future ideas around this kind of calculation (and my future interest in detecting "unexpected changes" without an event as opposed to corrective_change and also - detecting idempotency issues with resources automatically).

So I'm going to give it another go, and see what mess I end up with. You're a bastard btw. Just sayin' :-).